// @angular/fire/esm2022/vertexai-preview/vertexai.module.mjs
import { InjectionToken, Injector, NgModule, NgZone, Optional, makeEnvironmentProviders, } from '@angular/core';
import { VERSION, ɵAngularFireSchedulers, ɵgetDefaultInstanceOf } from '@angular/fire';
import { ɵAppCheckInstances } from '@angular/fire';
import { FirebaseApp, FirebaseApps } from '@angular/fire/app';
import { registerVersion } from 'firebase/app';
import { VERTEX_AI_PROVIDER_NAME, VertexAI, VertexAIInstances } from './vertexai';
import * as i0 from "@angular/core";
export const PROVIDED_VERTEX_AI_INSTANCES = new InjectionToken('angularfire2.vertexai-instances');
export function defaultVertexAIInstanceFactory(provided, defaultApp) {
    const defaultVertexAI = ɵgetDefaultInstanceOf(VERTEX_AI_PROVIDER_NAME, provided, defaultApp);
    return defaultVertexAI && new VertexAI(defaultVertexAI);
}
export function vertexAIInstanceFactory(fn) {
    return (zone, injector) => {
        const vertexAI = zone.runOutsideAngular(() => fn(injector));
        return new VertexAI(vertexAI);
    };
}
const VERTEX_AI_INSTANCES_PROVIDER = {
    provide: VertexAIInstances,
    deps: [
        [new Optional(), PROVIDED_VERTEX_AI_INSTANCES],
    ]
};
const DEFAULT_VERTEX_AI_INSTANCE_PROVIDER = {
    provide: VertexAI,
    useFactory: defaultVertexAIInstanceFactory,
    deps: [
        [new Optional(), PROVIDED_VERTEX_AI_INSTANCES],
        FirebaseApp,
    ]
};
export class VertexAIModule {
    constructor() {
        registerVersion('angularfire', VERSION.full, 'vertexai');
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: VertexAIModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.0.0", ngImport: i0, type: VertexAIModule });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: VertexAIModule, providers: [
            DEFAULT_VERTEX_AI_INSTANCE_PROVIDER,
            VERTEX_AI_INSTANCES_PROVIDER,
        ] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: VertexAIModule, decorators: [{
            type: NgModule,
            args: [{
                    providers: [
                        DEFAULT_VERTEX_AI_INSTANCE_PROVIDER,
                        VERTEX_AI_INSTANCES_PROVIDER,
                    ]
                }]
        }], ctorParameters: () => [] });
export function provideVertexAI(fn, ...deps) {
    registerVersion('angularfire', VERSION.full, 'vertexai');
    return makeEnvironmentProviders([
        DEFAULT_VERTEX_AI_INSTANCE_PROVIDER,
        VERTEX_AI_INSTANCES_PROVIDER,
        {
            provide: PROVIDED_VERTEX_AI_INSTANCES,
            useFactory: vertexAIInstanceFactory(fn),
            multi: true,
            deps: [
                NgZone,
                Injector,
                ɵAngularFireSchedulers,
                FirebaseApps,
                [new Optional(), ɵAppCheckInstances],
                ...deps,
            ]
        }
    ]);
}
